### PR TITLE
gh-116738: Make _abc module thread-safe

### PR DESCRIPTION
**Pull request by swtaarrs**
This is a collection of small changes aimed at making the `_abc` module safe to use in a free-threaded build:
- Only read and write `_abcmodule_state.abc_invalidation_counter` and `_abc_data._abc_negative_cache_version` with atomic operations (except in situations when the object should only be visible to a single thread, like initialization, teardown, or GC traverse).
- Change the two members above from `unsigned long long` to `uint64_t`. This was partially to avoid having to add more `_Py_atomic_*_ullong` variants, but also because `unsigned long long` is guaranteed to be at least 64 bits and I can't imagine we'd ever want more than 64. Might as well make it explicit.
- Change `_in_weak_set()` and `_add_to_weak_set()` to both take an `_abc_data *` and `PyObject **`, to allow them to use critical sections when reading or writing the pointers to sets of weak references. None of the `PyObject *`s that hold a set will change once they're first initialized, so we don't need to use locking when operating on the sets, only when reading or initializing the pointers.
- For the most part, no locks are held around multiple operations on related data structures (`_abc__abc_subclasscheck_impl()` being a good example). User code that does things like performing ABC subclass checks while concurrently registering subclasses will always be subject to surprising results no matter what we do internally, so the module only ensures that its data structures are kept internally consistent.
- Two functions modify a type's `tp_flags` member: `_abc__abc_init()` (should only be called with new types not visible to other threads) and `_abc__abc_register_impl()` (called with existing types). I added a helper to `typeobject.c` to support the `_abc_register` case, and it was just a few more lines to also support the `_abc_init` case as well. This felt a bit heavy-handed so I'm open to suggestions.

- Issue: gh-116738

### CODE REVIEW

File: Modules/_abc.c
Comment (colesbury) [inline Modules/_abc.c]: I think there is a missing `Py_DECREF` for the non-`PyLong_CheckExact` case. I would write this as:

```c
PyObject *flags;
if (PyDict_GetItemRef(dict, &_Py_ID(__abc_tpflags__), &flags) < 0) {
    return NULL;
}
if (flags == NULL || !PyLong_CheckExact(flags)) {
    Py_XDECREF(flags);
    Py_RETURN_NONE;
}

long val = PyLong_AsLong(flags);
Py_DECREF(flags);
...
```

File: Modules/_abc.c
Comment (colesbury) [inline Modules/_abc.c]: Let's just use `PyList_GetItemRef` unconditionally.

**Review by colesbury — COMMENTED** [review]
Overall looks good. A few comments below.

File: Modules/_abc.c
Comment (swtaarrs) [reply Modules/_abc.c]: I think that has one behavioral difference with the existing code: if `__abc_tpflags__` is present but not an `int`, the old code will still delete it from the type's dict, while your suggested code won't. Do you think it's worth preserving that?

Comment (colesbury) [reply Modules/_abc.c]: Oh, I missed that. Yeah, I think that's worth preserving.

Comment (swtaarrs) [reply Modules/_abc.c]: In that case, I think it's simpler on balance to keep the overall structure mostly the same and add the missing `Py_DECREF()`. I'll update the PR soon and you can let me know what you think.

Another option is to delete from the dict before checking if it's an int, but that would cause a behavioral difference in the case where the user provides a bad int value for `__abc_tpflags__`.

Comment (colesbury) [reply Modules/_abc.c]: Or just replace `PyDict_GetItemRef` with `PyDict_Pop`.

There were two motivations for my suggestion above:

1) Having `Py_DECREF()` spread across the code makes it hard to tell if we're properly reference counting.
2) I find the `flags_result == 1` or `flags_result == 0` types more confusing than something like `flags == NULL` because the meaning of 0 vs. 1 depends on the specific function you are calling (as opposed to `-1`, which consistently means error).

**Review by colesbury — APPROVED** [review]

**Comment by swtaarrs** [comment]
Don't merge this yet - I'm still going to make a small improvement to the refcounting of `flags` (hopefully after my next meeting)

**Comment by swtaarrs** [comment]
Ok, this should be good to go once the builds finish.

**Pull request merged**
